WikiWatershed/model-my-watershed
# Add Hydrology Totals

## Overview

Adds a footer element to the Hydrology table, and populates it with the total sum of each column.

## Demo

![mmw-hydrology-total](https://cloud.githubusercontent.com/assets/1430060/17572170/23daf7e2-5f21-11e6-9838-55beb1002924.gif)

## Testing Instructions

Checkout the branch and update packages to get the latest bootstrap table.

```bash
./scripts/npm.sh install
./scripts/bundle.sh --vendor
```

Load or create a MapShed project. Wait for the calculations to complete, then observe the Hydrology table. It should have a new "Total" row at the bottom, which should scroll horizontally with the rest of the table. Sorting the columns should not affect the row, which should stay at the bottom. The sum should be accurate for each column.

Ensure that other instances of bootstrap table (in the Analyze, Water Quality, and TR-55 tabs) continue to work as they should.

## Caveats

The bootstrap table's footer element is not well implemented, and has a number of slightly breaking edge cases. Resizing the window seems to disalign the columns ever so slightly. Overscrolling (on macOS or iOS for example) has the bounce effect only on the main table, not the header or the footer. Finally, sorting the columns triggers a recalculation of the totals which is why they flicker. I'm not quite sure how to address that.

Connects #1371
